File: server/src/main/java/com/fluffy/reaction/domain/LikeService.java
package com.fluffy.reaction.domain;

import com.fluffy.global.exception.BadRequestException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class LikeService {

    private final ReactionRepository reactionRepository;

    @Transactional
    public Long like(Like like, Long memberId) {
        String targetType = like.target().name();
        Long targetId = like.targetId();

        Reaction reaction = reactionRepository
                .findByTargetTypeAndTargetIdAndMemberIdAndType(targetType, targetId, memberId, ReactionType.LIKE)
                .orElseGet(() -> {
                    Reaction newReaction = new Reaction(targetType, targetId, memberId, ReactionType.LIKE);
                    return reactionRepository.save(newReaction);
                });

        reaction.active();

        return reaction.getId();
    }

    @Transactional
    public Long removeLike(Like like, Long memberId) {
        String targetType = like.target().name();
        Long targetId = like.targetId();

        Reaction reaction = reactionRepository
                .findByTargetTypeAndTargetIdAndMemberIdAndType(targetType, targetId, memberId, ReactionType.LIKE)
                .orElseThrow(() -> new BadRequestException("좋아요를 한 상태가 아닙니다."));

        reaction.delete();

        return reaction.getId();
    }
}

File: server/src/main/kotlin/com/fluffy/reaction/domain/LikeService.kt
package com.fluffy.reaction.domain

import com.fluffy.global.exception.NotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


@Service
class LikeService(
    private val reactionRepository: ReactionRepository
) {

    @Transactional
    fun like(like: Like, memberId: Long): Long {
        val targetType = like.target.name
        val targetId = like.targetId

        val reaction = reactionRepository
            .findByTargetTypeAndTargetIdAndMemberIdAndType(targetType, targetId, memberId, ReactionType.LIKE)
            ?: reactionRepository.save(Reaction.create(targetType, targetId, memberId, ReactionType.LIKE))

        reaction.activate()

        return reaction.id
    }

    @Transactional
    fun removeLike(like: Like, memberId: Long): Long {
        val targetType = like.target.name
        val targetId = like.targetId

        val reaction = reactionRepository
            .findByTargetTypeAndTargetIdAndMemberIdAndType(targetType, targetId, memberId, ReactionType.LIKE)
            ?: throw NotFoundException("아직 좋아요를 하지 않았습니다. 좋아요 정보: $like")

        reaction.delete()

        return reaction.id
    }
}
